tests/TestCase/ORM/QueryRegressionTest.php

```diff
@@ -1757,4 +1758,192 @@ public function testFormatDeepDistantAssociationRecords2()
         $this->assertEquals($expected, $query->toArray());
     }
+
+    /**
+     * Tests that subqueries can be used with function expressions.
+     *
+     * @return void
+     */
+    public function testFunctionExpressionWithSubquery()
+    {
+        $this->loadFixtures('Articles');
+        $table = $this->getTableLocator()->get('Articles');
+
+        $query = $table
+            ->find()
+            ->select(function (Query $q) use ($table) {
+                return [
+                    'value' => $q->func()->ABS([
+                        $table
+                            ->getConnection()
+                            ->newQuery()
+                            ->select(-1),
+                    ])
+                ];
+            });
+
+        $result = $query->first()->get('value');
+        $this->assertEquals(1, $result);
+    }
+
+    /**
+     * Tests that correlated subqueries can be used with function expressions.
+     *
+     * @return void
+     */
+    public function testFunctionExpressionWithCorrelatedSubquery()
+    {
+        $this->loadFixtures('Articles', 'Authors');
+        $table = $this->getTableLocator()->get('Articles');
+        $table->belongsTo('Authors');
+
+        $query = $table
+            ->find()
+            ->select(function (Query $q) use ($table) {
+                return [
+                    'value' => $q->func()->UPPER([
+                        $table
+                            ->getAssociation('Authors')
+                            ->find()
+                            ->select(['Authors.name'])
+                            ->where(function (QueryExpression $exp) {
+                                return $exp->equalFields('Authors.id', 'Articles.author_id');
+                            })
+                    ])
+                ];
+            });
+
+        $result = $query->first()->get('value');
+        $this->assertEquals('MARIANO', $result);
+    }
+
+    /**
+     * Tests that subqueries can be used with multi argument function expressions.
+     *
+     * @return void
+     */
+    public function testMultiArgumentFunctionExpressionWithSubquery()
+    {
+        $this->loadFixtures('Articles', 'Authors');
+        $table = $this->getTableLocator()->get('Articles');
+
+        $query = $table
+            ->find()
+            ->select(function (Query $q) use ($table) {
+                return [
+                    'value' => $q->func()->ROUND(
+                        [
+                            $table
+                                ->getConnection()
+                                ->newQuery()
+                                ->select(1.23456),
+                            2
+                        ],
+                        [null, 'integer']
+                    )
+                ];
+            });
+
+        $result = $query->first()->get('value');
+        $this->assertEquals('1.23', $result);
```

Comment:
This should be `assertSame()`.

There is regression here on `4.x` with SQL Server on AppVeyor.

https://ci.appveyor.com/project/cakephp/cakephp/builds/26914135#L296

Decimal in `3.x` is float and string in `4.x`, and `'1.23' == 1.23000`, but `'1.23' !== '1.23000'`

Reply:
I don't think there's any type conversion on CakePHP's part involved, given that `value` isn't a mapped column, both values are strings. I think it's PHPUnit doing a conversion somewhere, probably a loose check in earlier versions, and a strict one in later versions, ie something like `'1.23' == '1.230000'` vs `'1.23' === '1.230000'`.

Assigning a float return type to the function and comparing against a float should work I think.

Reply:
It was just my guess of failure reason, so you are probably right.

Anyway, this should be fixed in some way to pass AppVeyor 4.x build.